specification/draft-2/README.md

```diff
@@ -0,0 +1,687 @@
+# Core specification
+
+CWL specification comes in two parts:
+
+ - Core spec allows description of basic tools and workflows
+ - Extensions add features that enable describing more complex tools and workflows
+
+If using an extension feature, it needs to be explicitly listed in the requirements section
+of the description document. Vendors might also add features incompatible with rest of CWL -
+in this case, they should be explicitly listed as requirements with full URI as value of `class` property.
+
+
+## Command line tools
+
+Basic tool example:
+
+```yaml
+class: CommandLineTool
+label: Grep
+description: Filter lines of text file by regular expression.
+inputs:
+  - id: pattern
+    type: string
+    binding:
+      position: 1
+  - id: in_file
+    type: File
+    streamable: true
+    binding:
+      position: 2
+outputs:
+  - id: out_file
+    type: File
+    streamable: true
+    binding:
+      glob: output.txt
+baseCmd: grep
+stdout: output.txt
+successCodes: [0, 1]
+```
+
+Above is defined a command line tool labeled "grep" with two required inputs
+(a file and a string) and one file output.
+
+Input objects contain a `binding` section which describes how the arguments
+are passed onto the command line. In this case, `pattern` and `in_file` are
+positional arguments and `pattern` is passed first, followed by `in_file`.
+Identifiers for inputs and outputs must be unique across the document.
+
+The `binding` section of the output object contains a single property `glob`
+with a glob pattern describing how to recognize file generated by the tool.
+
+Additionally, `baseCmd` and `stdout` properties describe the executable name
+(optionally with additional hard-coded arguments like `python -m mymodule`)
+and standard output redirection (into a file called "output.txt").
+
+Success or failure of a job is determined by process exit code.
+By default, 0 is success and everything else is failure.
+However, some tools use exit codes to signal some information.
+The grep tool will exit with code 1 if no lines matched the pattern.
+For this purpose, we use the `successCodes` property to enumerate valid exit codes.
+
+Values for the `type` property can be any valid Apache Avro schema (in JSON form)
+with the `File` named type pre-defined. The `File` type declares that the input
+value should be passed in form of a file reference. At runtime, the file object
+itself will have at least the `path` and `size` properties, describing its
+local path and size in bytes.
+
+The File input and output of this tool are marked with the `streamable` flag to
+indicate that the tool will not seek over these files. CWL implementations might
+make use of this to pipe data from one tool to another.
+
+### Running tools
+
+A CWL implementation can run the above grep tool using a `Job` object like:
+
+```yaml
+inputs:
+  in_file:
+    __class: File
+    path: /local/path/to/file.txt
+    size: 42
+  pattern: find ?me
+allocatedResources:
+  cpu: 1
+  mem: 1024
+```
```

Comment:
Is __class required or just for disambigiouation?  Maybe it should be "@type": "cwl:File" ?

What if allocated resources is handled separately?  job.resources.yml alongside job.input.yml?  The allocated resources are sort of a pseudo-input, in some cases (like debugging a job) you want to know that two jobs are the same except for their resource allocation, in other cases (you just want to know if two jobs had the same input) the resource allocation is irrelevant.